src/fluentd/src/fluent-plugin-pgjson/lib/fluent/plugin/out_pgjson.rb

```diff
@@ -167,6 +167,40 @@ def insert_framework(hex_id, time, record)
       end
     end
 
+    def insert_task(hex_id, time, record)
+      # This function try to insert the task snapshot into task history table.
+      # In some cases, the framework controller may have duplicate logs about one task attempt,
+      # or there has been already successful inserted record before.
+      # To handle it, `insert_task` executes a "SELECT" first.
+      # If the uid exists in the table, it will ignore it safely.
+      # Any error should be raised.
+      thread = Thread.current
+      frameworkName = record["objectSnapshot"]["metadata"]["name"][0..31]
+      attemptIndex = record["objectSnapshot"]["metadata"]["annotations"]["FC_FRAMEWORK_ATTEMPT_ID"]
+      taskroleName = record["objectSnapshot"]["metadata"]["annotations"]["FC_TASKROLE_NAME"]
+      taskIndex = record["objectSnapshot"]["metadata"]["annotations"]["FC_TASK_INDEX"]
+      taskUid = record["objectSnapshot"]["metadata"]["uid"]
+      taskAttemptIndex = record["objectSnapshot"]["status"]["attemptStatus"]["id"]
+      podUid = record["objectSnapshot"]["status"]["attemptStatus"]["podUID"]
+      historyType = "retry"
```

Comment:
Note that podUid may be null

Reply:
I think it is OK, because the database column allows null. Could you give a real example so that I can test it?

Reply:
```
{
  "kind": "Task",
  "apiVersion": "frameworkcontroller.microsoft.com/v1",
  "metadata": {
    "name": "rsservices1-a-1",
    "namespace": "default",
    "uid": "e54be00f-feda-11ea-ac09-dc4a3e8c21dc",
    "resourceVersion": "28215906",
    "generation": 14,
    "creationTimestamp": "2020-09-25T02:57:49Z",
    "deletionTimestamp": "2020-09-25T02:59:16Z",
    "deletionGracePeriodSeconds": 0,
    "labels": {
      "FC_FRAMEWORK_NAME": "rsservices1",
      "FC_TASKROLE_NAME": "a",
      "FC_TASK_INDEX": "1"
    },
    "annotations": {
      "FC_CONFIGMAP_NAME": "rsservices1-attempt",
      "FC_CONFIGMAP_UID": "e5841754-feda-11ea-a80e-000d3af801a9",
      "FC_FRAMEWORK_ATTEMPT_ID": "0",
      "FC_FRAMEWORK_ATTEMPT_INSTANCE_UID": "0_e5841754-feda-11ea-a80e-000d3af801a9",
      "FC_FRAMEWORK_NAME": "rsservices1",
      "FC_FRAMEWORK_NAMESPACE": "default",
      "FC_FRAMEWORK_UID": "e53a17f6-feda-11ea-a80e-000d3af801a9",
      "FC_TASKROLE_NAME": "a",
      "FC_TASKROLE_UID": "e54b91e1-feda-11ea-ac08-dc4a3e8c21dc",
      "FC_TASK_INDEX": "1"
    }
  },
  "spec": null,
  "status": {
    "index": 1,
    "instanceUID": "e54be00f-feda-11ea-ac09-dc4a3e8c21dc",
    "startTime": "2020-09-25T02:57:49Z",
    "completionTime": "2020-09-25T02:59:16Z",
    "state": "Completed",
    "transitionTime": "2020-09-25T02:59:16Z",
    "deletionPending": true,
    "retryPolicyStatus": {
      "totalRetriedCount": 0,
      "accountableRetriedCount": 0,
      "retryDelaySec": null
    },
    "attemptStatus": {
      "id": 0,
      "startTime": "2020-09-25T02:57:49Z",
      "runTime": "2020-09-25T02:57:54Z",
      "completionTime": "2020-09-25T02:59:15Z",
      "instanceUID": "0_e61690cc-feda-11ea-a80e-000d3af801a9",
      "podName": "rsservices1-a-1",
      "podUID": "e61690cc-feda-11ea-a80e-000d3af801a9",
      "podNodeName": "minikube",
      "podIP": "172.17.0.7",
      "podHostIP": "10.169.8.232",
      "completionStatus": {
        "code": -230,
        "phrase": "DeleteTaskRequested",
        "type": {
          "name": "Failed",
          "attributes": [
            "Permanent"
          ]
        },
        "diagnostics": "User has requested to delete the Task by Framework ScaleDown"
      }
    }
  }
}
```

Reply:
Task.Spec can be nil if and only a TaskRole is directly deleted by scale down.